### PR TITLE
feat(infra): Provide logs in Azure Functions and Directory.Build.props

### PR DESCRIPTION
**Pull request by dechev**
- Logging in functions #96;
- Enable Nullable, ImplicitUsings and TreatWarningsAsErrors via Directory.Build.props;
- Fix all nullable-related warnings/errors

### CODE REVIEW

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs [hunk 1]
```diff
@@ -55,17 +55,23 @@ private static void WriteIntro(NewAzureDeprecationV1Message newNoticeV1MessageQu
 
         private static void WriteNotice(NewAzureDeprecationV1Message newNoticeV1MessageQueueMessage, StringBuilder issueBuilder)
         {
-            issueBuilder.AppendLine("### Notice");
-            issueBuilder.AppendLine();
-            issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.Notice.Description);
+            if (!string.IsNullOrWhiteSpace(newNoticeV1MessageQueueMessage.Notice?.Description))
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/IssueFactory.cs:58]: Can you revert this please?

We should always create this section. If there is no notice; we should add that in the 3rd append line

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs [hunk 2]
```diff
@@ -55,17 +55,23 @@ private static void WriteIntro(NewAzureDeprecationV1Message newNoticeV1MessageQu
...
+            }
         }
 
         private static void WriteImpact(NewAzureDeprecationV1Message newNoticeV1MessageQueueMessage, StringBuilder issueBuilder)
         {
-            issueBuilder.AppendLine("### Impact");
-            issueBuilder.AppendLine();
-            issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.Impact.Description);
-            issueBuilder.AppendLine();
+            if (!string.IsNullOrWhiteSpace(newNoticeV1MessageQueueMessage.Impact?.Description))
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/IssueFactory.cs:68]: Can you revert this please?

We should always create this section. If there is no description; we should add that in the 3rd append line

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs [hunk 3]
```diff
@@ -110,10 +116,13 @@ private static void WriteTimelineEntriesForAllPhases(List<InputTimeLineEntry> ti
         }
 
         private static void WriteRequiredAction(NewAzureDeprecationV1Message newNoticeV1MessageQueueMessage, StringBuilder issueBuilder)
-        {
-            issueBuilder.AppendLine("### Required Action");
-            issueBuilder.AppendLine();
-            issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.RequiredAction.Description);
+        {
+            if (!string.IsNullOrWhiteSpace(newNoticeV1MessageQueueMessage.RequiredAction?.Description))
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/IssueFactory.cs:120]: Can you revert this please?

We should always create this section. If there is no description; we should add that in the 3rd append line

File: src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs [hunk 4]
```diff
@@ -1,29 +1,35 @@
...
 using CloudNative.CloudEvents;
 using Microsoft.Azure.WebJobs;
 using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.Logging;
+using System.Net.Mime;
 
 namespace AzureDeprecation.Notifications.Functions
 {
-    public class EventGridNotificationsFunction
+    public partial class EventGridNotificationsFunction
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs:13]: Why are we making it partial?

File: src/AzureDeprecation.Notices.Management/Functions/NewAzureDeprecationNotificationFunction.cs [hunk 5]
```diff
@@ -1,49 +1,93 @@
...
 using Microsoft.Azure.WebJobs;
-using Microsoft.Azure.WebJobs.ServiceBus;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
 using SBEntityType = Microsoft.Azure.WebJobs.ServiceBus.ServiceBusEntityType;
 
 namespace AzureDeprecation.Notices.Management.Functions
 {
-    public class NewAzureDeprecationNotificationFunction
+    public partial class NewAzureDeprecationNotificationFunction
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/Functions/NewAzureDeprecationNotificationFunction.cs:12]: Why are we making it partial?

File: src/AzureDeprecation.Runtimes.AzureFunctions/ValueStopwatch.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+﻿using System.Diagnostics;
+
+namespace AzureDeprecation.Runtimes.AzureFunctions
+{
+    public struct ValueStopwatch
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Runtimes.AzureFunctions/ValueStopwatch.cs:5]: What is this used for? I didn't see it so must have looked over it

File: src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs [hunk 7]
```diff
@@ -32,16 +38,23 @@ public async Task Run([ServiceBusTrigger("new-deprecation-notices", "event-grid-
                 .UsingAuthenticationKey(eventGridAuthKey)
                 .Build();
             
-            var @event = new CloudEvent(CloudEventsSpecVersion.V1_0,
+            var @event = new CloudEvent(
+                CloudEventsSpecVersion.V1_0,
                 "NewDeprecationNoticePublishedV1",
-                new Uri("https://github.com/azure-deprecation/dashboard"),
-                subject: $"/{newDeprecationNoticePublishedV1Message.DeprecationInfo.Impact.Services.First()}")
+                new Uri("https://github.com/azure-deprecation/dashboard"), // TODO: dev/staging dashboard
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs:44]: Meh, I think this is OK :D but we can make it configurable if you want

File: src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs [hunk 8]
```diff
@@ -84,19 +80,22 @@ private List<string> DetermineRequiredLabels(NewAzureDeprecationV1Message newNot
                 "verified"
             };
 
-            foreach (var service in newNoticeV1MessageQueueMessage.Impact.Services)
-            {
-                var label = LabelFactory.GetForService(service);
-                labels.Add(label);
+            if (newNoticeV1MessageQueueMessage.Impact is not null)
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs:83]: If it's not there, we should add "unknown" labels

File: src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs (hunk 4 above)
Comment (dechev) [reply src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs:13]: For the `LoggerMessage`'s further down to work: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/loggermessage?view=aspnetcore-6.0

File: src/AzureDeprecation.Runtimes.AzureFunctions/ValueStopwatch.cs (hunk 6 above)
Comment (dechev) [reply src/AzureDeprecation.Runtimes.AzureFunctions/ValueStopwatch.cs:5]: Shamelessly copied (and extended for the milliseconds) from here: https://github.com/dotnet/aspnetcore/blob/main/src/Shared/ValueStopwatch/ValueStopwatch.cs

File: src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs (hunk 7 above)
Comment (dechev) [reply src/AzureDeprecation.Notifications/Functions/EventGridNotificationsFunction.cs:44]: Not immediately maybe, just saw that we'll be linking **the** dashboard from whatever db this ends up being stored in.

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs (hunk 1 above)
Comment (dechev) [reply src/AzureDeprecation.Notices.Management/IssueFactory.cs:60]: Adjusted, but with TODO for proper wording

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs (hunk 2 above)
Comment (dechev) [reply src/AzureDeprecation.Notices.Management/IssueFactory.cs:74]: Adjusted, but with TODO for proper wording

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs (hunk 3 above)
Comment (dechev) [reply src/AzureDeprecation.Notices.Management/IssueFactory.cs:130]: Adjusted, but with TODO for proper wording

File: src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs (hunk 8 above)
Comment (dechev) [reply src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs:83]: Can you please check if I got what you asked for?

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs [hunk 9]
```diff
@@ -55,17 +55,31 @@ private static void WriteIntro(NewAzureDeprecationV1Message newNoticeV1MessageQu
...
-            issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.Notice.Description);
+            issueBuilder.AppendLine("### Notice");
+            issueBuilder.AppendLine();
+            if (!string.IsNullOrWhiteSpace(newNoticeV1MessageQueueMessage.Notice?.Description))
+            {
+                issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.Notice.Description);
+            }
+            else
+            {
+                issueBuilder.AppendLine(); // TODO: come up with proper wording for missing Notice.Description
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/IssueFactory.cs:66]:
```suggestion
                issueBuilder.AppendLine("No notice information is available.");

```

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs [hunk 10]
```diff
@@ -55,17 +55,31 @@ private static void WriteIntro(NewAzureDeprecationV1Message newNoticeV1MessageQu
...
-            issueBuilder.AppendLine();
+            issueBuilder.AppendLine("### Impact");
+            issueBuilder.AppendLine();
+            if (!string.IsNullOrWhiteSpace(newNoticeV1MessageQueueMessage.Impact?.Description))
+            {
+                issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.Impact.Description);
+            }
+            else
+            {
+                issueBuilder.AppendLine(); // TODO: come up with proper wording for missing Impact.Description
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/IssueFactory.cs:80]:
```suggestion
                issueBuilder.AppendLine("No impact information is available.");

```

File: src/AzureDeprecation.Notices.Management/IssueFactory.cs [hunk 11]
```diff
@@ -110,10 +124,17 @@ private static void WriteTimelineEntriesForAllPhases(List<InputTimeLineEntry> ti
...
+        {
+            issueBuilder.AppendLine("### Required Action");
+            issueBuilder.AppendLine();
+            if (!string.IsNullOrWhiteSpace(newNoticeV1MessageQueueMessage.RequiredAction?.Description))
+            {
+                issueBuilder.AppendLine(newNoticeV1MessageQueueMessage.RequiredAction.Description);
+            }
+            else
+            {
+                issueBuilder.AppendLine(); // TODO: come up with proper wording for missing RequiredAction.Description
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/IssueFactory.cs:136]:
```suggestion
                issueBuilder.AppendLine("No required action information is available.");

```

File: src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs (hunk 8 above)
Comment (tomkerkhove) [reply src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs:83]: The idea is to rather use `services:unknown`, `area:unknown`, etc - Added suggestion below

File: src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs [hunk 12]
```diff
@@ -84,18 +80,25 @@ private List<string> DetermineRequiredLabels(NewAzureDeprecationV1Message newNot
...
+                var areaLabel = LabelFactory.GetForImpactArea(newNoticeV1MessageQueueMessage.Impact.Area);
+                labels.Add(areaLabel);
+                var typeLabel = LabelFactory.GetForImpactType(newNoticeV1MessageQueueMessage.Impact.Type);
+                labels.Add(typeLabel);
+                var cloudLabel = LabelFactory.GetForCloud(newNoticeV1MessageQueueMessage.Impact.Cloud);
+                labels.Add(cloudLabel);
+            }
+            else
+            {
+                labels.Add("unknown");
```
Comment (tomkerkhove) [inline src/AzureDeprecation.Notices.Management/MessageHandlers/NewAzureDeprecationV1MessageMessageHandler.cs:100]:
```suggestion
                labels.Add("services:unknown");
                labels.Add("area:unknown");
                labels.Add("cloud:unknown");
                labels.Add("impact:unknown");
```

**Review by tomkerkhove — APPROVED** [review]
LGTM - Added some final comments but looks good, thank you.

**Pull request merged**